File: src/rtweekendpkg/vec3.nim
import std/math
import std/strformat

type Vec3* = object
  e: array[3, float]

proc newVec3*: Vec3 =
  Vec3(
    e: [0.0, 0.0, 0.0]
  )

proc newVec3*(x: float, y: float, z: float): Vec3 =
  Vec3(
    e: [x, y, z]
  )

proc x*(v: Vec3): float =
  v.e[0]

proc y*(v: Vec3): float =
  v.e[1]

proc z*(v: Vec3): float =
  v.e[2]

proc `-`*(v: Vec3): Vec3 =
  newVec3(-v.x, -v.y, -v.z)

proc `[]`*(v: Vec3, i: int): float =
  v.e[i]

proc `[]`*(v: var Vec3, i: int): var float =
  v.e[i]

proc `+=`*(v: var Vec3, u: Vec3): var Vec3 =
  for i in 0..<3:
    v.e[i] += u.e[i]
  v

proc `*=`*(v: var Vec3, t: float): var Vec3 =
  for i in 0..<3:
    v.e[i] *= t
  v

proc `/=`*(v: var Vec3, t: float): var Vec3 =
  v *= 1 / t

proc lengthSquared*(v: Vec3): float =
  v.x * v.x + v.y * v.y + v.z * v.z

proc length*(v: Vec3): float =
  sqrt(v.lengthSquared)

proc `$`*(v: Vec3): string =
  fmt"{v.x} {v.y} {v.z}"

proc `+`*(u: Vec3, v: Vec3): Vec3 =
  newVec3(u.x + v.x, u.y + v.y, u.z + v.z)

proc `-`*(u: Vec3, v: Vec3): Vec3 =
  newVec3(u.x - v.x, u.y - v.y, u.z - v.z)

proc `*`*(v: Vec3, t: float): Vec3 =
  newVec3(v.x * t, v.y * t, v.z * t)

proc `*`*(t: float, v: Vec3): Vec3 =
  v * t

proc `/`*(v: Vec3, t: float): Vec3 =
  (1 / t) * v

proc dot*(u: Vec3, v: Vec3): float =
  u.x * v.x + u.y * v.y + u.z * v.z

proc cross*(u: Vec3, v: Vec3): Vec3 =
  newVec3(
    u.y * v.z - u.z * v.y,
    u.z * v.x - u.x * v.z,
    u.x * v.y - u.y * v.x
  )

proc unitVector*(v: Vec3): Vec3 =
  v / v.length

type Point3* = Vec3

proc newPoint3*(x: float, y: float, z: float): Point3 =
  newVec3(x, y, z)

type Color* = Vec3

proc newColor*(x: float, y: float, z: float): Color =
  newVec3(x, y, z)

File: src/Vec3.zig
const std = @import("std");

e: [3]Component,

const Component = f64;
const Self = @This();

pub fn init(x: Component, y: Component, z: Component) Self {
    return .{ .e = .{ x, y, z } };
}

pub fn neg(self: Self) Self {
    return init(-self.e[0], -self.e[1], -self.e[2]);
}

pub fn get(self: Self, i: usize) Component {
    return self.e[i];
}

pub fn getMut(self: *Self, i: usize) *Component {
    return &self.e[i];
}

pub fn iadd(self: *Self, v: Self) *Self {
    self.e[0] += v.e[0];
    self.e[1] += v.e[1];
    self.e[2] += v.e[2];
    return self;
}

pub fn iscale(self: *Self, t: Component) *Self {
    self.e[0] *= t;
    self.e[1] *= t;
    self.e[2] *= t;
    return self;
}

pub fn iscaleInv(self: *Self, t: Component) *Self {
    self.e[0] /= t;
    self.e[1] /= t;
    self.e[2] /= t;
    return self;
}

pub fn length(self: Self) Component {
    return std.math.sqrt(self.lengthSquared());
}

pub fn lengthSquared(self: Self) Component {
    return self.e[0] * self.e[0] +
        self.e[1] * self.e[1] +
        self.e[2] * self.e[2];
}

pub fn format(self: Self, comptime _: []const u8, _: std.fmt.FormatOptions, writer: anytype) !void {
    try writer.print("{d} {d} {d}", .{ self.e[0], self.e[1], self.e[2] });
}

pub fn add(u: Self, v: Self) Self {
    return .{ .e = .{
        u.e[0] + v.e[0],
        u.e[1] + v.e[1],
        u.e[2] + v.e[2],
    } };
}

pub fn sub(u: Self, v: Self) Self {
    return .{ .e = .{
        u.e[0] - v.e[0],
        u.e[1] - v.e[1],
        u.e[2] - v.e[2],
    } };
}

pub fn componentMul(u: Self, v: Self) Self {
    return .{ .e = .{
        u.e[0] * v.e[0],
        u.e[1] * v.e[1],
        u.e[2] * v.e[2],
    } };
}

pub fn scale(v: Self, t: Component) Self {
    return .{ .e = .{
        v.e[0] * t,
        v.e[1] * t,
        v.e[2] * t,
    } };
}

pub fn scaleInv(v: Self, t: Component) Self {
    return .{ .e = .{
        v.e[0] / t,
        v.e[1] / t,
        v.e[2] / t,
    } };
}

pub fn dot(u: Self, v: Self) Component {
    return u.e[0] * v.e[0] +
        u.e[1] * v.e[1] +
        u.e[2] * v.e[2];
}

pub fn cross(u: Self, v: Self) Self {
    return .{ .e = .{
        u.e[1] * v.e[2] - u.e[2] * v.e[1],
        u.e[2] * v.e[0] - u.e[0] * v.e[2],
        u.e[0] * v.e[1] - u.e[1] * v.e[0],
    } };
}

pub fn unit(v: Self) Self {
    return v.scaleInv(v.length());
}
